### DISCUSSION TITLE
Settings for Module Extensions

### DISCUSSION THREAD

**Discussion by duydvu** [directus/directus#18019] [Feature Requests]
### Changelog
| Date  | Description |
| ------------- | ------------- |
| Apr 4 | Update `directus_modules` schema  |

### Summary

Hi Directus team,
I would like to request a new feature that allows modules to have their own setting options. Similar to how we can configure options such as placeholder or font for interfaces.

Here is a suggestion for a UI change:
In Project Settings > Modules, each module has an edit button next to the enable button, click the button to configure the module.
![Screenshot 2023-04-01 215432](https://user-images.githubusercontent.com/23261928/229296908-16b7e0c3-5828-440d-afb4-0536e3f991bb.png)

The options for a module are defined in the `index.ts` file.

### Motivation

While interface, layout, panel & display extensions can be configured very easily, there is no native method to configure module extensions.

For instance, we have an idea for the [Schema Management Module Extension](https://github.com/rezo-labs/directus-extension-schema-management-module) that allow importing schema presets from a configurable source, similar to how Docker can be configured to use another public/private Docker Registry. At the moment, the source URL for schema presets is hard-coded so if we can dynamically set it, anyone could host their own presets.

Besides this, configurable modules add a lot of possibilities. For example:
- Select which roles can access the module by combining with `preRegisterCheck`.
- Enable/disable certain features/routes.
- Pass settings to other extensions (interface, display, layout). These extensions can behave differently in each module.
- "Clone" modules. For example: Duplicate the Content module so that each one shows a subset of collections. This will keep the navigation bar clean as the number of collections grows.

### Detailed Design

Here is my idea to implement this feature, please correct me if there is something wrong.

1. Add `options` to `defineModule` to define the schema for module options.
2. New System Collection `directus_modules`:
    - **id**: integer - ID of the module.
    - **type**: string - module or link.
    - **module_id**: string - correspond to the `id` in `defineModule` config.
    - **name**: string - Name of the module. Used to override the one in `defineModule`.
    - **icon**: string - Same as name.
    - **url**: string - Used when type="link".
    - **options**: json - contain options set by admin.
    - **enabled**: boolean - enable the module if true.
    - **locked**: boolean - if true, the module cannot be disabled.
    - **order**: integer - the order in which the module appears in the module bar.
    - Many items in this table can have the same `module_id` so a module can be duplicated.
3. New APIs:
    - GET /modules - get all modules.
    - POST /modules - create a new module.
    - PATCH /modules/:id - update a module.
    - DELETE /modules/:id - delete a module.
4. `<module-bar />` component uses a new store called `modulesStore` to know which module to show.
5. During app hydration, call /modules to get all modules and save in `modulesStore.modules`.
6. Pass options to the corresponding module.
7. Update UI in the Module setting section.

### Must Have:
- Configure module settings in Project Settings > Modules.
- Define options in `index.ts`.

### Should Have:
- Duplicate modules.

### Could Have:
- Add `visibleCollections` and `roles` options to the Content module.

### Won't Have:
- N/A

### Drawbacks

I see that the development of Directus has been slowed recently so I'm willing to help if you don't have enough resources to implement this feature. Just check my design above.

### Alternatives

N/A

### Adoption Strategy

To deploy this feature to existing Directus projects, first create module items in `directus_modules` during initialization, using the values in `settingsStore.settings.module_bar`.
It will only need to run once so we need to check the number of items in `directus_modules`, if > 0 then don't create.
The `settingsStore.settings.module_bar` will become obsolete so we can remove it later.

**Comment by rijkvanzanten** [comment]
Question! Should there be a way to have those options per user as well? Options for layouts for examples are saved to directus_presets rather than a "global admin configured" layouts config 🤔 

> I see that the development of Directus has been slowed recently

See https://github.com/directus/directus/discussions/17977 for ideas on how to fix that 😄

**Comment by duydvu** [comment]
Yeah, there should be both options for global & per user. They can be merged before passing to the module component.
Some options should only be updated by admin (imagine a user can change the roles so they can access a forbidden module) and prevent user options to override them.

With this in mind, the updated design should be:
- Add `user` & `role` to `directus_modules`.
- Unlike layouts, we must obtain both the global & user options and merge them before passing.
- In `defineModule`, have a new option `adminOptions` (string[]) to prevent using user options.

And where to place the user options?
1. Option 1: The right sidebar, like layouts.
2. Option 2: In User Directory module. (I like this one personally because the options panel UI is consistent with the admin one)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
